### DISCUSSION TITLE
SPI:  MSBFIRST/LSBFIRST and BigEndian/LittleEndian are independent settings

### DISCUSSION THREAD

**Discussion by board707** [earlephilhower/arduino-pico#1508] [General]
Hi
There is some confusion between these two pairs of settings, but as far I know they are generally independent. The MSBFIRST/LSBFIRST controls the order of bits in a single byte, the BigEndian/LittleEndian - the order of bytes in a 32-bit word. Changing the order of bits when setting the LSBFIRST does not mean changing the order of bytes.
However, in your code these settings are treated as related. For example, in the `reverse16Bit` procedure, when changing from MSB to LSB, not only bits, but also bytes are unfolded:
```
inline uint16_t SPIClassRP2040::reverse16Bit(uint16_t w) {
    return (reverseByte(w & 0xff) << 8) | (reverseByte(w >> 8));
}
```

**Comment by earlephilhower** [comment]
The Arduino docs are pretty high-level, so do you have any other core/MCU you use that does the MSB/LSB switch differently?  I'd like to just confirm, especially since this call is only used when users explicitly try and read 16 bits (which to me implies those 16-bits are a logical value and we should do magic to make it usable as a `short int` here in the core).  But, that's just my thinking in absence of other evidence...

Would you like to make a PR fixing this, since you found it (and I assume you ran into a problem because of it)?

**Comment by board707** [comment]
Thank for your answer
I haven't run into a problem yet, however I'm writing code and I need to provide to users the ability to change LSB to MSB. And I'm not sure exactly how that should work.
As I wrote, there is some confusion in this matter. These two settings are described to be independent. For example, I work with the Winner Micro controller W806, which does not change the order of bits in bytes when changing the order of bytes from LITTLE to BIGENDIAN. But I understand that this can hardly be evidence.
I'm sorry, it would probably be more correct to transform this to discussion - but I don't see such an option. My post is not a complaint about your package, but rather a reflection on whether it works correctly.

**Comment by earlephilhower** [comment]
Converted to discussion for you.  Never heard of the W806 and only can seem to find a machine translated overview (i.e. not register map/etc.) from the GOOG.  But, I will see what, say, the old-school AVRs do, as that's easily discovered and really the OG Arduino definition. :)

**Comment by earlephilhower** [comment]
According to this AVR core, `transfer16` should do byte swapping, too:

https://github.com/arduino/ArduinoCore-avr/blob/master/libraries/SPI/src/SPI.h#L220-L241

So, I think the existing code is behaving correctly.

**Reply by board707** [reply]
Thank you for your answer, but examples of code can hardly serve as a sure proof. 
By definition, these two settings should be independent.
I will continue to study the issue. I would like to find some documentation
thanks for the help

**Reply by earlephilhower** [reply]
While I agree scientifically with you that
> examples of code can hardly serve as a sure proof.

Practically, though, where there is no formal spec and the existing code is ~ 20 years old and written by the keepers of the (unwritten) spec...like in the Arduino.cc case, I would say it's as much proof as is possible...
